OsmAnd/src/net/osmand/plus/helpers/WaypointHelper.java

```diff
@@ -244,29 +244,27 @@ public AlarmInfo getMostImportantAlarm(SpeedConstants sc, boolean showCameras) {
 				LocationPointWrapper lwp = lp.get(kIterator);
 				AlarmInfo inf = (AlarmInfo) lwp.point;
 				int currentRoute = route.getCurrentRoute();
-				if (inf.getLocationIndex() < currentRoute && inf.getLastLocationIndex() != -1
-						&& inf.getLastLocationIndex() < currentRoute) {
-					// skip
+				// getLastLocationIndex()  == -1 is always < currentRoute
+				if (inf.getLocationIndex() <= currentRoute && inf.getLastLocationIndex() < currentRoute) {
+					// skip already passed alarms
 				} else {
-					if (inf.getType() == AlarmInfoType.TUNNEL && inf.getLastLocationIndex() != -1
-							&& currentRoute > inf.getLocationIndex()
-							&& currentRoute < inf.getLastLocationIndex()) {
-						inf.setFloatValue(route.getDistanceToPoint(inf.getLastLocationIndex()));
-					}
+					int distanceByRoute = 0;
 					Location lastKnownLocation = app.getRoutingHelper().getLastProjection();
-					int d = (int) Math.max(0.0, MapUtils.getDistance(lastKnownLocation.getLatitude(), lastKnownLocation.getLongitude(),
-							inf.getLatitude(), inf.getLongitude()) - lwp.getDeviationDistance());
-					if (inf.getLocationIndex() == currentRoute && d > 10) {
-						return null;
+					if (inf.getLocationIndex() < currentRoute) {
+						// update remaining length
+						inf.setFloatValue(route.getDistanceToPoint(lastKnownLocation, inf.getLastLocationIndex()));
+					} else {
+						distanceByRoute = route.getDistanceToPoint(lastKnownLocation, inf.getLocationIndex() - 1);
 					}
-					if (!atd.isTurnStateActive(0, d, STATE_LONG_PNT_APPROACH)) {
+					if (!atd.isTurnStateActive(0, distanceByRoute, STATE_SHORT_PNT_APPROACH)) { // TODO or STATE_LONG_PNT_APPROACH
```

Comment:
Todo not ok in result code

Reply:
If to change STATE_LONG_PNT_APPROACH to the STATE_SHORT_PNT_APPROACH, will appear issue  https://github.com/osmandapp/OsmAnd/issues/3772